apps/pushes/views/diff.py

```diff
@@ -179,10 +179,9 @@ def real_rev(self, rev):
                    .filter(branch=1)  # default branch
                    .order_by('-pk')
                    .values_list('revision', flat=True)[0])
-        # Convert the 'from' and 'to' to strings (instead of unicode)
-        # in case mercurial needs to look for the key in binary data.
-        # This prevents UnicodeWarning messages.
-        ctx = self.client[str(rev)]
+        # Convert the 'from' and 'to' to bytestrings
+        # which is what hglib expexts.
+        ctx = self.client[rev.encode('ascii', 'replace')]
```

Comment:
I wonder about using `replace` here. Can that not lead to silent errors? If `rev` can be a user input, then it could lead to an erroneous value, but what you will see here will be an index error instead of a decoding error. I might be misleading during debug.

If I'm misunderstanding, can you tell me why you're using `replace` here and in some other places?

Reply:
Yeah, I was trying to deal with people pushing bad data into the diff view. As they have bad intents, let's just error early.

Updated the patch to do so, and add tests for it.